pkg/prober/prober.go

```diff
@@ -0,0 +1,158 @@
+package prober
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/component"
+	"github.com/prometheus/common/route"
+)
+
+const (
+	healthyEndpointPath  = "/-/healthy"
+	readyEndpointPath    = "/-/ready"
+	okProbeText          = "thanos %v is %v"
+	errorProbeText       = "thanos %v is not %v. Reason: %v"
+	probeErrorHTTPStatus = 503
+	initialErrorText     = "thanos %s is initializing"
+)
+
+// Prober represents health and readriness status of given compoent.
+type Prober struct {
+	logger       log.Logger
+	componentMtx sync.Mutex
+	component    component.Component
+	readyMtx     sync.Mutex
+	readiness    error
+	healthyMtx   sync.Mutex
+	healthiness  error
+}
+
+// SetLogger sets logger used by the Prober.
+func (p *Prober) SetLogger(logger log.Logger) {
+	p.logger = logger
+}
+
+func (p *Prober) getLogger() log.Logger {
+	return p.logger
+}
+
+// SetComponent sets component name of the Prober displayed in responses.
+func (p *Prober) SetComponent(component component.Component) {
+	p.componentMtx.Lock()
+	defer p.componentMtx.Unlock()
+	p.component = component
+}
+
+func (p *Prober) getComponent() component.Component {
+	p.componentMtx.Lock()
+	defer p.componentMtx.Unlock()
+	return p.component
+}
+
+// NewProber returns Prober representing readiness and healthiness of given component.
+func NewProber(component component.Component, logger log.Logger) *Prober {
+	initialErr := fmt.Errorf(initialErrorText, component)
+	prober := &Prober{}
+	prober.SetComponent(component)
```

Comment:
Prober should not use public method TBH (:

Reply:
Can we just create prober with filled struct?

Reply:
Definitely, just out of curiosity why should be this avoided? Is that generic thing in Golang not to use own public methods?

Reply:
I removed also calls to `SetNotReady` and `SetNotHealthy` which caused that the mutex locking is gone but since it's the instantiation of the struct it should be ok I hope?

Reply:
Because there is no need for this. Literally we will never in our code require dynamic reload of component. We don't have components that transforms into another components right? 

Every bit of genercity is introducing complexity. Since we don't need that now (YAGNI rule), we can avoid that complexity, right? (:

Reply:
Makes sense, thanks for for the explanation :)